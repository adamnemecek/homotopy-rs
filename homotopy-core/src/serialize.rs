use crate::util::FastHashMap;
use crate::{rewrite::Cone, Cospan, Diagram, DiagramN, Generator, Rewrite, Rewrite0, RewriteN};
use highway::{HighwayHash, HighwayHasher};
use serde::{Deserialize, Serialize};
use std::{
    collections::BTreeMap, convert::TryInto, hash::Hash, marker::PhantomData, num::NonZeroU32,
};

/// Similar to `Hash`, except supposed to be deterministic and shouldn't collide
trait Keyed<K> {
    fn key(&self) -> K;
}

impl<K, H: Hash> Keyed<Key<K>> for H {
    fn key(&self) -> Key<K> {
        let mut h = HighwayHasher::default();
        self.hash(&mut h);
        let hash = h.finalize128();
        (u128::from(hash[1]) + (u128::from(hash[0]) << 64)).into()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Store {
    #[serde(skip_serializing, skip_deserializing)]
    diagram_keys: FastHashMap<Diagram, Key<Diagram>>,
    diagrams: BTreeMap<Key<Diagram>, DiagramSer>,

    #[serde(skip_serializing, skip_deserializing)]
    rewrite_keys: FastHashMap<Rewrite, Key<Rewrite>>,
    rewrites: BTreeMap<Key<Rewrite>, RewriteSer>,

    #[serde(skip_serializing, skip_deserializing)]
    cone_keys: FastHashMap<Cone, Key<Cone>>,
    cones: BTreeMap<Key<Cone>, ConeSer>,
}

impl Store {
    pub fn new() -> Self {
        Self {
            diagram_keys: Default::default(),
            diagrams: Default::default(),
            rewrite_keys: Default::default(),
            rewrites: Default::default(),
            cone_keys: Default::default(),
            cones: Default::default(),
        }
    }

    pub fn pack_diagram(&mut self, diagram: &Diagram) -> Key<Diagram> {
        if let Some(key) = self.diagram_keys.get(diagram) {
            return *key;
        }

        let serialized = match diagram {
            Diagram::Diagram0(generator) => DiagramSer::D0 {
                dimension: ZeroU32::default(),
                generator: *generator,
            },
            Diagram::DiagramN(diagram) => {
                let source = self.pack_diagram(&diagram.source());
                let cospans = diagram
                    .cospans()
                    .iter()
                    .map(|cospan| self.pack_cospan(cospan))
                    .collect();
                DiagramSer::Dn {
                    dimension: (diagram.dimension() as u32).try_into().unwrap(),
                    source,
                    cospans,
                }
            }
        };

        let key: Key<Diagram> = serialized.key();
        self.diagram_keys.insert(diagram.clone(), key);
        self.diagrams.insert(key, serialized);
        key
    }

    fn pack_cospan(&mut self, cospan: &Cospan) -> CospanSer {
        let forward = self.pack_rewrite(&cospan.forward);
        let backward = self.pack_rewrite(&cospan.backward);
        CospanSer { forward, backward }
    }

    pub fn pack_rewrite(&mut self, rewrite: &Rewrite) -> Key<Rewrite> {
        if let Some(key) = self.rewrite_keys.get(&rewrite) {
            return *key;
        }

        let serialized = match rewrite {
            Rewrite::Rewrite0(Rewrite0(None)) => RewriteSer::R0 {
                dimension: ZeroU32::default(),
                source: None,
                target: None,
            },
            Rewrite::Rewrite0(Rewrite0(Some((x, y)))) => RewriteSer::R0 {
                dimension: ZeroU32::default(),
                source: Some(*x),
                target: Some(*y),
            },
            Rewrite::RewriteN(rewrite) => {
                let cones = rewrite
                    .cones()
                    .iter()
                    .map(|cone| self.pack_cone(cone))
                    .collect();
                RewriteSer::Rn {
                    dimension: (rewrite.dimension() as u32).try_into().unwrap(),
                    cones,
                }
            }
        };

        let key: Key<Rewrite> = serialized.key();
        self.rewrite_keys.insert(rewrite.clone(), key);
        self.rewrites.insert(key, serialized);
        key
    }

    fn pack_cone(&mut self, cone: &Cone) -> ConeWithIndexSer {
        if let Some(key) = self.cone_keys.get(cone) {
            return ConeWithIndexSer {
                index: cone.index as u32,
                cone: *key,
            };
        }

        let serialized = ConeSer {
            source: cone
                .internal
                .source
                .iter()
                .map(|cospan| self.pack_cospan(cospan))
                .collect(),
            target: self.pack_cospan(&cone.internal.target),
            slices: {
                cone.internal
                    .slices
                    .iter()
                    .map(|slice| self.pack_rewrite(slice))
                    .collect()
            },
        };

        let key: Key<Cone> = serialized.key();
        self.cone_keys.insert(cone.clone(), key);
        self.cones.insert(key, serialized);
        ConeWithIndexSer {
            index: cone.index as u32,
            cone: key,
        }
    }

    pub fn unpack_diagram(&self, key: Key<Diagram>) -> Option<Diagram> {
        match self.diagrams.get(&key)?.clone() {
            DiagramSer::D0 { generator, .. } => Some(Diagram::from(generator)),
            DiagramSer::Dn {
                source, cospans, ..
            } => {
                let source = self.unpack_diagram(source)?;
                let cospans = cospans
                    .into_iter()
                    .map(|cospan| self.unpack_cospan(&cospan))
                    .collect::<Option<_>>()?;
                Some(DiagramN::new_unsafe(source, cospans).into())
            }
        }
    }

    fn unpack_cospan(&self, serialized: &CospanSer) -> Option<Cospan> {
        let forward = self.unpack_rewrite(serialized.forward)?;
        let backward = self.unpack_rewrite(serialized.backward)?;
        Some(Cospan { forward, backward })
    }

    pub fn unpack_rewrite(&self, key: Key<Rewrite>) -> Option<Rewrite> {
        match self.rewrites.get(&key)?.clone() {
            RewriteSer::R0 { source, target, .. } => match (source, target) {
                (None, None) => Some(Rewrite0(None).into()),
                (Some(source), Some(target)) => Some(Rewrite0(Some((source, target))).into()),
                (None, Some(_)) | (Some(_), None) => None,
            },
            RewriteSer::Rn { dimension, cones } => {
                let cones = cones
                    .into_iter()
                    .map(|cone| self.unpack_cone(cone))
                    .collect::<Option<_>>()?;
                Some(RewriteN::new(u32::from(dimension) as usize, cones).into())
            }
        }
    }

    fn unpack_cone(&self, cone: ConeWithIndexSer) -> Option<Cone> {
        let serialized = self.cones.get(&cone.cone)?.clone();
        let source = serialized
            .source
            .into_iter()
            .map(|cospan| self.unpack_cospan(&cospan))
            .collect::<Option<_>>()?;
        let target = self.unpack_cospan(&serialized.target)?;
        let slices = serialized
            .slices
            .into_iter()
            .map(|slice| self.unpack_rewrite(slice))
            .collect::<Option<_>>()?;
        Some(Cone::new(cone.index as usize, source, target, slices))
    }
}

impl Default for Store {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
#[serde(untagged)]
enum DiagramSer {
    D0 {
        dimension: ZeroU32,
        generator: Generator,
    },
    Dn {
        dimension: NonZeroU32,
        source: Key<Diagram>,
        cospans: Vec<CospanSer>,
    },
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
#[serde(untagged)]
enum RewriteSer {
    R0 {
        dimension: ZeroU32,
        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(default)]
        source: Option<Generator>,
        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(default)]
        target: Option<Generator>,
    },
    Rn {
        dimension: NonZeroU32,
        cones: Vec<ConeWithIndexSer>,
    },
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
struct CospanSer {
    forward: Key<Rewrite>,
    backward: Key<Rewrite>,
}

#[derive(Debug, PartialEq, Copy, Eq, Hash, Clone, Serialize, Deserialize)]
struct ConeWithIndexSer {
    index: u32,
    cone: Key<Cone>,
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
struct ConeSer {
    source: Vec<CospanSer>,
    target: CospanSer,
    slices: Vec<Key<Rewrite>>,
}

// Phantom key type
#[derive(Debug)]
pub struct Key<K>(u128, PhantomData<K>);

impl<K> Serialize for Key<K> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&format!("{:032x}", self.0))
    }
}

impl<'de, K> Deserialize<'de> for Key<K> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct HexVisitor;

        impl<'de> serde::de::Visitor<'de> for HexVisitor {
            type Value = u128;

            fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(f, "a hex encoded key")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                u128::from_str_radix(v, 16).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_string(HexVisitor).map(Self::from)
    }
}

impl<K> From<u128> for Key<K> {
    fn from(k: u128) -> Self {
        Self(k, PhantomData)
    }
}

impl<K> Clone for Key<K> {
    fn clone(&self) -> Self {
        Self(self.0, PhantomData)
    }
}

impl<K> Copy for Key<K> {}

impl<K> PartialEq for Key<K> {
    fn eq(&self, other: &Self) -> bool {
        self.0.eq(&other.0)
    }
}

impl<K> Eq for Key<K> {}

impl<K> PartialOrd for Key<K> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl<K> Ord for Key<K> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.0.cmp(&other.0)
    }
}

impl<K> Hash for Key<K> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default)]
struct ZeroU32;

impl Serialize for ZeroU32 {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_u32(0)
    }
}

impl<'de> Deserialize<'de> for ZeroU32 {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let number = u32::deserialize(deserializer)?;

        if 0 == number {
            Ok(ZeroU32)
        } else {
            Err(serde::de::Error::invalid_value(
                serde::de::Unexpected::Unsigned(u64::from(number)),
                &"zero",
            ))
        }
    }
}

#[allow(clippy::derive_hash_xor_eq)]
impl Hash for ZeroU32 {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        state.write_u32(0);
    }
}
