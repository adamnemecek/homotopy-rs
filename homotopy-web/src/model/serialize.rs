use super::{AttachOption, Color, GeneratorInfo, Signature, Workspace};
use homotopy_core::common::{Generator, SliceIndex};
use homotopy_core::serialize::{Key, Keyed, Serialization};
use homotopy_core::Diagram;
use im::{HashMap, Vector};
use wasm_bindgen::JsCast;

pub fn generate_download(name: &str, data: &[u8]) -> Result<(), wasm_bindgen::JsValue> {
    let val: js_sys::Uint8Array = data.into();
    let mut options = web_sys::BlobPropertyBag::new();
    options.type_("application/msgpack");
    let blob = web_sys::Blob::new_with_u8_array_sequence_and_options(
        &js_sys::Array::of1(&val.into()).into(),
        &options,
    )?;
    let url = web_sys::Url::create_object_url_with_blob(&blob)?;
    let window = web_sys::window().ok_or("no window")?;
    let document = window.document().ok_or("no document")?;
    let body = document.body().ok_or("no body")?;
    let e = document.create_element("a")?;
    let a = e
        .dyn_ref::<web_sys::HtmlElement>()
        .ok_or("failed to create anchor")?;
    a.set_attribute("href", &url)?;
    a.set_attribute("download", &name)?;
    body.append_child(&a)?;
    a.click();
    a.remove();
    web_sys::Url::revoke_object_url(&url)
}

#[derive(Debug, PartialEq, Eq, Clone, Default, serde::Serialize, serde::Deserialize)]
pub(crate) struct Data {
    signature: Serialization,
    generator_info: HashMap<Generator, (String, Color)>,
    workspace: Option<WorkspaceSer>,
}

#[derive(Debug, PartialEq, Eq, Clone, serde::Serialize, serde::Deserialize)]
struct WorkspaceSer {
    diagram: Key<Diagram>,
    path: Vector<SliceIndex>,
    attach: Option<Vector<AttachOption>>,
    highlight: Option<AttachOption>,
}

impl From<Signature> for Data {
    fn from(sig: Signature) -> Self {
        let mut stripped: std::collections::HashMap<Generator, Diagram> = Default::default();
        let mut generator_info: HashMap<Generator, (String, Color)> = Default::default();
        for (k, v) in sig {
            stripped.insert(k, v.diagram);
            generator_info.insert(k, (v.name, v.color));
        }
        Self {
            signature: Serialization::from(stripped),
            generator_info,
            workspace: None,
        }
    }
}

impl From<(Signature, Workspace)> for Data {
    fn from((sig, ws): (Signature, Workspace)) -> Self {
        let mut stripped: std::collections::HashMap<Generator, Diagram> = Default::default();
        let mut generator_info: HashMap<Generator, (String, Color)> = Default::default();
        let k = ws.diagram.key();
        for (k, v) in sig {
            stripped.insert(k, v.diagram);
            generator_info.insert(k, (v.name, v.color));
        }
        Self {
            signature: Serialization::from((stripped, ws.diagram)),
            generator_info,
            workspace: Some(WorkspaceSer {
                diagram: k,
                path: ws.path,
                attach: ws.attach,
                highlight: ws.highlight,
            }),
        }
    }
}

impl From<Data> for (Signature, Option<Workspace>) {
    fn from(data: Data) -> Self {
        let signature = &data.signature;
        let workspace: Option<Workspace> = data.workspace.map(|ws| Workspace {
            diagram: signature.diagram(&ws.diagram),
            path: ws.path,
            attach: ws.attach,
            highlight: ws.highlight,
        });
        let sig: std::collections::HashMap<Generator, Diagram> = data.signature.into();
        let gi = data.generator_info;
        (
            sig.into_iter()
                .map(|(k, v)| {
                    (
                        k,
                        GeneratorInfo {
                            name: gi[&k].0.clone(),
                            color: gi[&k].1.clone(),
                            diagram: v,
                        },
                    )
                })
                .collect(),
            workspace,
        )
    }
}

impl From<Data> for Vec<u8> {
    fn from(data: Data) -> Self {
        rmp_serde::to_vec(&data).unwrap()
    }
}

impl From<Vec<u8>> for Data {
    fn from(bs: Vec<u8>) -> Self {
        rmp_serde::from_read_ref(&bs).unwrap()
    }
}
