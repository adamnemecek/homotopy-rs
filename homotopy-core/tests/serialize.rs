use homotopy_core::serialize::*;
use homotopy_core::*;
use insta::*;

#[test]
fn serialize_associator() {
    let (_, diagram) = examples::associator();

    let (serialized, key) = {
        let mut store = Store::new();
        let key = store.pack_diagram(&diagram.clone().into());
        let serialized = serde_json::to_string(&store).unwrap();
        (serialized, key)
    };

    assert_debug_snapshot!(serialized);

    let deserialized = {
        let store: Store = serde_json::from_str(&serialized).unwrap();
        store.unpack_diagram(key).unwrap()
    };

    assert_eq!(Diagram::from(diagram), deserialized);
}
