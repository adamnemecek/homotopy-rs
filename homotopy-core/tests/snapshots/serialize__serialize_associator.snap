---
source: homotopy-core/tests/serialize.rs
expression: serialized

---
"{\"diagrams\":{\"1f21266c5feda29ad12d57c4d994ec6b\":{\"dimension\":2,\"source\":\"89fff112fea69305c054fdb8dfaffcec\",\"cospans\":[{\"forward\":\"0e8c0dc8f9fa8e6530df35ead1e4961b\",\"backward\":\"9d739512a3b0f00ae59bd78f729b0d40\"},{\"forward\":\"0e8c0dc8f9fa8e6530df35ead1e4961b\",\"backward\":\"9d739512a3b0f00ae59bd78f729b0d40\"}]},\"89fff112fea69305c054fdb8dfaffcec\":{\"dimension\":1,\"source\":\"e3887b7ceb2031c6242aa3a697ad28de\",\"cospans\":[{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"},{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"},{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"}]},\"c71a9a9b9aa5ad1f1c2dd7c5ad96ed12\":{\"dimension\":3,\"source\":\"1f21266c5feda29ad12d57c4d994ec6b\",\"cospans\":[{\"forward\":\"1c1937c4466947457c100ec5ab3562a1\",\"backward\":\"f34df3f5d3609c718947b20ce4f77258\"}]},\"e3887b7ceb2031c6242aa3a697ad28de\":{\"dimension\":0,\"generator\":{\"dimension\":0,\"id\":0}}},\"rewrites\":{\"0e34c25b3a964bde3f1e36929270fb3d\":{\"dimension\":0,\"source\":{\"dimension\":0,\"id\":0},\"target\":{\"dimension\":3,\"id\":3}},\"0e8c0dc8f9fa8e6530df35ead1e4961b\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"1f64a05e25afb448a6235ccbf8c9f2d8\"}]},\"1c1937c4466947457c100ec5ab3562a1\":{\"dimension\":2,\"cones\":[{\"index\":0,\"cone\":\"87fdf117daee2b6bc6fe5595ade45542\"}]},\"2d0d59b33ec99b89dbbaa4a8b454fd60\":{\"dimension\":0,\"source\":{\"dimension\":1,\"id\":1},\"target\":{\"dimension\":3,\"id\":3}},\"474c05c50f16fbc32e2f61dae00570f7\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"f4b1417547252262e1f0767a113b6ae4\"}]},\"4c9b3f039211f19587c1e4af0e9df947\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"892a9891562312e2816ea32914c458da\"}]},\"6d30ba3752e25d523f6560724d8c82e9\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"e8f7899509e601dd7041d8dd345c48b1\"}]},\"716cd7a6b2d91e3244bcc99a5571dcca\":{\"dimension\":1,\"cones\":[{\"index\":1,\"cone\":\"1f64a05e25afb448a6235ccbf8c9f2d8\"}]},\"757248a846f2bdbe795d70bed6cd9217\":{\"dimension\":0,\"source\":{\"dimension\":0,\"id\":0},\"target\":{\"dimension\":2,\"id\":2}},\"8410e3366b89781e7bb5599bdea9bbac\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"d3f4db25030e89f2d4b54e3992d36eae\"}]},\"85269645a4beab137c0474e3c96e7ab2\":{\"dimension\":0,\"source\":{\"dimension\":0,\"id\":0},\"target\":{\"dimension\":1,\"id\":1}},\"87efd305047e717815db3b38187bbde0\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"2c8aec2277fc7fc243395623ac4724e0\"}]},\"8d0f2b4df15e7efb2a90c27b538ef01a\":{\"dimension\":0,\"source\":{\"dimension\":1,\"id\":1},\"target\":{\"dimension\":2,\"id\":2}},\"9d739512a3b0f00ae59bd78f729b0d40\":{\"dimension\":1,\"cones\":[{\"index\":0,\"cone\":\"4cbcacc16607d4664d069e6c1d0e76f7\"}]},\"c2ddc8323e0aa80bbd439832873badcc\":{\"dimension\":1,\"cones\":[{\"index\":1,\"cone\":\"4cbcacc16607d4664d069e6c1d0e76f7\"}]},\"dfaf64e54bf7f1a96771ea991f3e102c\":{\"dimension\":0,\"source\":{\"dimension\":2,\"id\":2},\"target\":{\"dimension\":3,\"id\":3}},\"f34df3f5d3609c718947b20ce4f77258\":{\"dimension\":2,\"cones\":[{\"index\":0,\"cone\":\"ed7a78e14123b5950caa19e2b514298f\"}]}},\"cones\":{\"1f64a05e25afb448a6235ccbf8c9f2d8\":{\"source\":[{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"},{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"}],\"target\":{\"forward\":\"757248a846f2bdbe795d70bed6cd9217\",\"backward\":\"757248a846f2bdbe795d70bed6cd9217\"},\"slices\":[\"8d0f2b4df15e7efb2a90c27b538ef01a\",\"8d0f2b4df15e7efb2a90c27b538ef01a\"]},\"2c8aec2277fc7fc243395623ac4724e0\":{\"source\":[{\"forward\":\"757248a846f2bdbe795d70bed6cd9217\",\"backward\":\"757248a846f2bdbe795d70bed6cd9217\"}],\"target\":{\"forward\":\"0e34c25b3a964bde3f1e36929270fb3d\",\"backward\":\"0e34c25b3a964bde3f1e36929270fb3d\"},\"slices\":[\"dfaf64e54bf7f1a96771ea991f3e102c\"]},\"4cbcacc16607d4664d069e6c1d0e76f7\":{\"source\":[{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"}],\"target\":{\"forward\":\"757248a846f2bdbe795d70bed6cd9217\",\"backward\":\"757248a846f2bdbe795d70bed6cd9217\"},\"slices\":[\"8d0f2b4df15e7efb2a90c27b538ef01a\"]},\"87fdf117daee2b6bc6fe5595ade45542\":{\"source\":[{\"forward\":\"0e8c0dc8f9fa8e6530df35ead1e4961b\",\"backward\":\"9d739512a3b0f00ae59bd78f729b0d40\"},{\"forward\":\"0e8c0dc8f9fa8e6530df35ead1e4961b\",\"backward\":\"9d739512a3b0f00ae59bd78f729b0d40\"}],\"target\":{\"forward\":\"4c9b3f039211f19587c1e4af0e9df947\",\"backward\":\"6d30ba3752e25d523f6560724d8c82e9\"},\"slices\":[\"8410e3366b89781e7bb5599bdea9bbac\",\"87efd305047e717815db3b38187bbde0\"]},\"892a9891562312e2816ea32914c458da\":{\"source\":[{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"},{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"},{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"}],\"target\":{\"forward\":\"0e34c25b3a964bde3f1e36929270fb3d\",\"backward\":\"0e34c25b3a964bde3f1e36929270fb3d\"},\"slices\":[\"2d0d59b33ec99b89dbbaa4a8b454fd60\",\"2d0d59b33ec99b89dbbaa4a8b454fd60\",\"2d0d59b33ec99b89dbbaa4a8b454fd60\"]},\"d3f4db25030e89f2d4b54e3992d36eae\":{\"source\":[{\"forward\":\"757248a846f2bdbe795d70bed6cd9217\",\"backward\":\"757248a846f2bdbe795d70bed6cd9217\"},{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"}],\"target\":{\"forward\":\"0e34c25b3a964bde3f1e36929270fb3d\",\"backward\":\"0e34c25b3a964bde3f1e36929270fb3d\"},\"slices\":[\"dfaf64e54bf7f1a96771ea991f3e102c\",\"2d0d59b33ec99b89dbbaa4a8b454fd60\"]},\"e8f7899509e601dd7041d8dd345c48b1\":{\"source\":[{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"}],\"target\":{\"forward\":\"0e34c25b3a964bde3f1e36929270fb3d\",\"backward\":\"0e34c25b3a964bde3f1e36929270fb3d\"},\"slices\":[\"2d0d59b33ec99b89dbbaa4a8b454fd60\"]},\"ed7a78e14123b5950caa19e2b514298f\":{\"source\":[{\"forward\":\"716cd7a6b2d91e3244bcc99a5571dcca\",\"backward\":\"c2ddc8323e0aa80bbd439832873badcc\"},{\"forward\":\"0e8c0dc8f9fa8e6530df35ead1e4961b\",\"backward\":\"9d739512a3b0f00ae59bd78f729b0d40\"}],\"target\":{\"forward\":\"4c9b3f039211f19587c1e4af0e9df947\",\"backward\":\"6d30ba3752e25d523f6560724d8c82e9\"},\"slices\":[\"474c05c50f16fbc32e2f61dae00570f7\",\"87efd305047e717815db3b38187bbde0\"]},\"f4b1417547252262e1f0767a113b6ae4\":{\"source\":[{\"forward\":\"85269645a4beab137c0474e3c96e7ab2\",\"backward\":\"85269645a4beab137c0474e3c96e7ab2\"},{\"forward\":\"757248a846f2bdbe795d70bed6cd9217\",\"backward\":\"757248a846f2bdbe795d70bed6cd9217\"}],\"target\":{\"forward\":\"0e34c25b3a964bde3f1e36929270fb3d\",\"backward\":\"0e34c25b3a964bde3f1e36929270fb3d\"},\"slices\":[\"2d0d59b33ec99b89dbbaa4a8b454fd60\",\"dfaf64e54bf7f1a96771ea991f3e102c\"]}}}"
